text/0000-declarative-shadow-dom-polyfill.md

```diff
@@ -0,0 +1,346 @@
+  <figcaption align="center">
+    <b>Fig.2: Performance comparison in Safari</b>
+  </figcaption>
+</figure>
+<br></br>
+ 
+ 
+**FIREFOX**
+
+<figure>
+  <img src="../assets/declarative-shadow-dom-polyfill/perf-firefox.png" alt="Performance comparison in Mozilla Firefox">
+  <figcaption align="center">
+    <b>Fig.3: Performance comparison in Mozilla Firefox</b>
+  </figcaption>
+</figure>
+
+## LWC and Declarative ShadowDOM polyfills
+
+From the three polyfills described in this RFC, LWC currently supports without any modification the [Single script shadow root attachment](#single-loop-script-shadow-root-attachment-reference) and the [Single MutationObserver](#single-mutationobserver-based-reference) polyfill. In order to support the [inline script polyfill](#inline-script-shadow-root-attachment-reference) approach, the application container will need to modify the SSR output, or LWC to provide a mechanism to modify the SSR output to include the script after the template.
+
+This RFC proposes that **there will be no modification** to the existing API of `@lwc/engine-server` to support the inline script polyfill approach.
+
+The proposed solution to application containers that want to use the inline-script approach, is to do string manipulation to add the script after the template tag of the SSR output. Example:
+
+
+```js
+import { renderComponent as lwcRenderComponent} from '@lwc/engine-server';
+
+const polyfill = `<script type="text/javascript">replaceTemplateWithShadowRoot(document.currentScript);</script>`;
+
+export function renderComponent(ceName, ctor, props) {
+   const originalSSROutput = lwcRenderComponent(ceName, ctor, props);
+   /* renders
+    <x-foo>
+        <template shadowroot="open">
+            <div class="foo">
+                <x-btn class="x-btn_btn-host">
+                    <template shadowroot="open">
+                        <button class="foo x-btn_btn">click me</button>
+                    </template>
+                </x-btn>
+            </div>
+        </template>
+    </x-foo>
+   */
+   
+   return originalSSROutput.replaceAll("</template>", `</template>${polyfill}`);
+   /* returns:
+    <x-foo>
+        <template shadowroot="open">
+            <div class="foo">
+                <x-btn class="x-btn_btn-host">
+                    <template shadowroot="open">
+                        <button class="foo x-btn_btn">click me</button>
+                    </template><script type="text/javascript">replaceTemplateWithShadowRoot(document.currentScript);</script>
+                </x-btn>
+            </div>
+        </template><script type="text/javascript">replaceTemplateWithShadowRoot(document.currentScript);</script>
+    </x-foo>
+    */
+}
+```
+
+Other alternative solutions were taken into account, but at the moment there’s not a strong use case for them, and other than removing the extra string manipulation operation, there’s no added benefit.
```

Comment:
Based on the benchmarks and the tradeoffs, this sounds reasonable to me. Assuming there is no other way for a `<template>` to sneak into the SSR output, the best approach seems to be to just ask container to do regex replacement. Otherwise we are offering a lot of API surface to maintain, which is not really much better than regex replacement.

Reply:
I'm also in favor of leaving the decision up to containers. It's up to them to decide and weigh the tradeoffs of Chromium vs other browsers, streaming vs end-to-end performance, etc.